Documentation: initial commit - CONTRIBUTING.md

JustWaveThings opened the pull request:
This PR creates a draft CONTRIBUTING.md from which the team can add to and change as needed. 

It is a direct result of the discussion here - https://github.com/orgs/Club-40/discussions/17

Note that there are several TODO's that link to issue templates that I think would be good to implement.

crespire commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,165 @@
+* [Issues and Bugs](#issue)
+* [Feature Requests](#feature)
+
+## <a name="question"></a> Got a Question?
```
Does `<a name="">` do the same thing as `<a id="question" href="#"></a>`?

Not a blocker, just curious. Otherwise looks good to me.

JustWaveThings replied:
Yep. same functionality. I just tested it.

crespire reviewed (approved):
Looks good to me, the comment I left is not a blocker.

justEstif commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,165 @@
+
+1. [Fork](https://github.com/Club-40/group-project-1 ) and then clone your fork.
+
+1. Load in the clone of your fork via the plugin loader of your choice.
```
I think you want to say `code editor` instead of `plugin loader`.

justEstif replied:
Why plugin loader? I think you want to say code editor.

justEstif commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,165 @@
+
+1. Make the required updates to the code.
+
+1. Push the new changes to your repository (this will update your Pull Request as well)
```
The list number should be changed from 1 to 2.

justEstif commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,165 @@
+    git checkout -b my-branch master
+    ```
+
+1. Make your changes, including tests when appropriate. Please refer to [`test/README.md`](TODO ) for a guide.
````
The list numbers need to be fixed (it repeats 1 for a couple of them)

crespire replied:
I think Github Markdown should auto convert the `1.` in groups into a list, though I am not sure if it will in this case with the extra line breaks in between.

justEstif replied:
Good catch. It looks right in the preview too.

justEstif reviewed (commented):
Left some comments mostly related to list numbers.

Lofty-Brambles commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+   You wouldn't want to duplicate existing efforts.
+2. Be sure that an issue describes the problem you're fixing, or documents the design for the feature you're adding.
+   Discussing the design upfront helps to ensure that we're ready to accept your work.
+3. [Fork](https://github.com/Club-40/group-project-1 ) and then clone your fork.
```
```suggestion
3. [Fork](https://github.com/Club-40/group-project-1) and then clone your fork.
```

Lofty-Brambles commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+
+5. Make your changes, including tests when appropriate. Please refer to [`test/README.md`](TODO ) for a guide.
+
+6. Follow the current coding style as close as possible. Simply take a look at other files and stick to their styling.
```
```suggestion
6. Follow the current coding style as closely as possible, in accordance to the provided style guidelines.
```

Lofty-Brambles commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,155 @@
+```txt
+<scope>: <short summary>
+   |           |
+   |           └─> Summary in present tense. Not capitalized. No period at the end.
````
```suggestion
   |           └─> Summary in the present tense. Not capitalized. No period at the end.
```

Lofty-Brambles commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+
+Explain the motivation for the change in the commit message body.
+This commit message should explain _why_ you are making the change.
+You can include a comparison of the previous behavior with the new behavior in order to illustrate the impact of the change.
```
```suggestion
You can include a comparison of the previous behavior with the new behavior to illustrate the impact of the change.
```

Lofty-Brambles commented on CONTRIBUTING.md:
I might be being picky about this one, but I'd rather prefer if you linted it to not add a newline at 80 or so characters but let the auto-wrapping work, just like the TOP curriculum repository.

Lofty-Brambles commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+
+1. Make the required updates to the code.
+
+1. Push the new changes to your repository (this will update your Pull Request as well)
```
```suggestion
2. Push the new changes to your repository (this will update your Pull Request as well)
```

Ananth-Adhikarla replied:
should be making a new branch and cloning the main branch instead of forking it.
Forking is mostly done for open source contributions where you are not a member of the organization ( teams )
unless you guys have a different idea that I am not aware of or this is for the general public.

crespire commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+<blank line>
+Fixes #<issue number>
+<blank line>
+Co-authored-by: name <name@example.com>
```
If you pair with someone, it's nice to have them included this way. It looks like this: https://github.com/rubyforgood/casa/commit/541ba846b651bc69eec0454b9468b84ece91c8dc

Lofty-Brambles replied:
Hmm, sure then.

Lofty-Brambles commented:
Oh yes, it's definitely something that can be done without forking. That part of the contributing guide is mainly for non-member peeps.

JustWaveThings commented:
Documentation: Add CONTRIBUTING.md with guidelines for submitting contributions
```diff
diff --git a/CONTRIBUTING.md b/CONTRIBUTING.md
index 89cdbc1..d3700c6 100644
--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,6 +1,24 @@
 # DRAFT Contributing Guidelines
 
-Thank you for considering a contribution to the club-40 group project!
+Thank you for considering a contribution to the [club-40](https://github.com/club-40) group project!
+
+## Table of Contents
+
+- [Got a Question?](#got-a-question)
+- [Found a Bug?](#found-a-bug)
+- [Missing a feature?](#missing-a-feature)
+- [Submission Guidelines](#submission-guidelines)
+  - [Submitting an Issue](#submitting-an-issue)
+  - [Submitting a Pull Request (PR)](#submitting-a-pull-request-pr)
+    - [Addressing review feedback](#addressing-review-feedback)
+    - [After your pull request is merged](#after-your-pull-request-is-merged)
+- [Commit Message Format](#commit-message-format)
+- [Commit Message Header](#commit-message-header)
+  - [Scope](#scope)
+  - [Summary](#summary)
+  - [Commit Message Body](#commit-message-body)
+  - [Commit Message Footer](#commit-message-footer)
+    - [Revert commits](#revert-commits)
 
 ## Got a Question?
 
@@ -17,11 +35,11 @@ Or, better, submit a pull request with a fix readily available!
 You can request a new feature by submitting an issue to the issue tracker.
 If you would like to implement a new feature, please consider the impact it would have on the codebase:
 
-* For a major impact affecting multiple files, adding wholly new topics and sections to the documentation, etc.:
+- For a major impact affecting multiple files, adding wholly new topics and sections to the documentation, etc.:
     please open an issue first, outlining your proposal so that it can be discussed.
     This allows us to better coordinate our resources around your proposal and help you in crafting your contribution.
 
-* Small features can be put together and directly submitted as a Pull Request.
+- Small features can be put together and directly submitted as a Pull Request.
 
 ## Submission Guidelines
 
@@ -45,12 +63,12 @@ Before you submit your Pull Request, follow these steps:
 4. In your cloned repository, make your changes in a new branch:
 
     ```shell
-    git checkout -b my-branch master
+    git checkout -b my-branch dev
     ```
 
 5. Make your changes, including tests when appropriate. Please refer to [`test/README.md`](TODO ) for a guide.
 
-6. Follow the current coding style as close as possible. Simply take a look at other files and stick to their styling.
+6. Follow the current coding style as closely as possible, in accordance to the provided style guidelines.
 
 7. Commit your changes using a descriptive commit message following our [commit message conventions](TODO).
 
@@ -68,7 +86,7 @@ If we ask for changes via code reviews then:
 
 1. Make the required updates to the code.
 
-1. Push the new changes to your repository (this will update your Pull Request as well)
+2. Push the new changes to your repository (this will update your Pull Request as well)
 
    ```shell
    git push --force-with-lease
@@ -102,8 +120,7 @@ The `footer` is optional. The commit footer format describes what the footer is
 
 ```txt
 <scope>: <short summary>
-   |           |
-   |           └─> Summary in present tense. Not capitalized. No period at the end.
+   |           └─> Summary in the present tense. Not capitalized. No period at the end.
    |
    └─> Commit Scope: Where did this commit happen?
          e.g. docs, syntax, folding, keybinds, github, legal, ...
@@ -118,9 +135,9 @@ individual pull requests should they not depend on each other.
 
 Use the summary field to provide a succinct description of the change:
 
-* use the imperative, present tense: "change" not "changed" nor "changes"
-* don't capitalize the first letter
-* no dot (.) at the end
+- use the imperative, present tense: "change" not "changed" nor "changes"
+- don't capitalize the first letter
+- no dot (.) at the end
 
 ### Commit Message Body
 
@@ -128,7 +145,7 @@ Just as in the summary, use the imperative, present tense: "fix" not "fixed" nor
 
 Explain the motivation for the change in the commit message body.
 This commit message should explain _why_ you are making the change.
-You can include a comparison of the previous behavior with the new behavior in order to illustrate the impact of the change.
+You can include a comparison of the previous behavior with the new behavior to illustrate the impact of the change.
 
 ### Commit Message Footer
 
@@ -151,5 +168,5 @@ If the commit reverts a previous commit, it should begin with `revert:`, followe
 
 The content of the commit message body should contain:
 
-* information about the SHA of the commit being reverted in the following format: `This reverts commit <SHA>`,
-* a clear description of the reason for reverting the commit message.
+- information about the SHA of the commit being reverted in the following format: `This reverts commit <SHA>`,
+- a clear description of the reason for reverting the commit message.

```
